core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -6184,13 +6184,16 @@ private void processPartitionRemoveCandidates() {
     private void processPartitionRemoveCandidates0(int n) {
         boolean anyReadersBeforeCommittedTxn = checkScoreboardHasReadersBeforeLastCommittedTxn();
         // This flag will determine to schedule O3PartitionPurgeJob at the end or all done already.
-        boolean scheduleAsyncPurge = anyReadersBeforeCommittedTxn;
+        boolean scheduleAsyncPurge = false;
+        long lastCommittedTxn = this.getTxn();
 
-        if (!anyReadersBeforeCommittedTxn) {
-            for (int i = 0; i < n; i += 2) {
-                try {
-                    final long timestamp = partitionRemoveCandidates.getQuick(i);
-                    final long txn = partitionRemoveCandidates.getQuick(i + 1);
+        for (int i = 0; i < n; i += 2) {
+            try {
+                final long timestamp = partitionRemoveCandidates.getQuick(i);
+                final long txn = partitionRemoveCandidates.getQuick(i + 1);
+                // txn >= lastCommittedTxn means there are some versions found in the table directory
+                // that are not attached to the table most likely as result of a rollback
+                if (!anyReadersBeforeCommittedTxn || txn >= lastCommittedTxn) {
```

Comment:
do you know under what circumstances the `unlinkOrRemove()` below may fail? 

I assume it's more common on Windows when someone else has the file open? If that happens then we are still prone to the original bug, aren't we?

Reply:
we should not commit any new transaction before the rollbacked txn dir is safely removed. is that correct?
can we somehow guarantee it?

Reply:
On Windows it can easily fail when the files are open. Noting should open the files before the transaction is committed, but the writer can potentially not clean itself fully on a rollback and keep some files open but this is only a potential scenario.
I don't believe we should fail to commit in a rare case when we can not remove a directory here. I think it's an overkill given that 
- it does not immediately lead to disaster (there are more conditions like open reader at the specific transaction, purge job run at a particular time). 
- the reading error is not terminal, e.g. after retrying the query, it will reload the reader to the latest partition directories and go through